### ISSUE TITLE
Support Integer in the config.toml via the "env()" function

### ISSUE THREAD

**Issue by jones1008** [supabase/cli#1551]
**Is your feature request related to a problem? Please describe.**
In my `config.toml` I tried to setup my ports to reference environment variables as follows:
```toml
[api]
port = "env(SUPABASE_API_PORT)"
```

```dotenv
SUPABASE_API_PORT=1234
```

When trying to run `supabase start` now it throws me the following error:
```
Error: cannot read config in /home/jfl/projects/poodle/poodle-frontend: toml: line 7 (last key "api.port"): incompatible types: TOML value has type string; destination has type integer
Have you set up the project with supabase init?
```

**Describe the solution you'd like**
It seems the `env()` function [is only supported for strings](https://supabase.com/docs/guides/cli/managing-config#using-secrets-inside-configtoml) inside the `config.toml`. It would be nice to also be able to configure your ports (type Integer) via environment variables

We would need a solution that is `.toml` syntax compliant and allows for integer values (and potentially other types. I don't know the implementation of how supabase-cli replaces the `env()` with its variables at runtime (I don't know Go), but maybe we could make something work like follows:
```toml
[api]
port = "env(SUPABASE_API_PORT) as integer"
```
The supabase-cli then could remove the whole string (including `"`) so we would have only the integer at runtime but still have a valid `.toml` syntax.

**Describe alternatives you've considered**
I tried to omit the `"` around the `env()` but that is a .toml syntax error:
```toml
[api]
port = env(SUPABASE_API_PORT)
```
Error message:
```
$ supabase start
Error: cannot read config in /home/jfl/projects/poodle/poodle-frontend: toml: line 7 (last key "api.port"): expected value but found "env" instead
```

**Additional context**
-

**Comment by thenbe** [comment]
Unable to set the port through an environment variable is painful.

The current workaround for this issue is to manually duplicate these values (e.g. `$SUPABASE_API_PORT` now needs to be explicitly declared both in your `.env` and in `config.toml`.

This is an issue because it leaves the `config.toml` vulnerable to going out-of-sync. (e.g. a team member updates the value of `SUPABASE_API_PORT` in `.env` but does not update it in `config.toml`).

### Cause

`config.toml` already knows how to interpolate from environment variables. But it fails when the _config key expects a integer_. If the config key expects a string, everything works fine.

```toml
# config.toml

# this works (because `s3_region` expects value of type string)
s3_region = "env(S3_REGION)"

# this does not work (because `port` expects value of type integer)
port = "env(SUPABASE_API_PORT)"
```

```
toml: line 7 (last key "api.port"): incompatible types: TOML value has type string; destination has type integer
```

### Proposal

For config keys that expect type integer, we should attempt to coerce the input into an integer before failing.

It's intuitive; this is the behavior users are going to naturally expect. And I can't think of any downsides to doing this.

**Comment by rileytaylor** [comment]
Being able to do this with boolean values as well would also solve issues in CI environments as discussed here: https://github.com/supabase/edge-runtime/issues/235

Being able to use `config.toml` as a driver for a 12-factor style env approach for the local environment and deploy workflows for 1+ hosted environments would _greatly_ simplify the environment workflow recommended in the docs too: https://supabase.com/docs/guides/cli/managing-environments

**Comment by jumski** [comment]
Having abilty to set ports in config.toml via env vars would help so much in current parallelize-work-via-worktreees-heavy landscape of ai assisted coding

**Comment by sweatybridge** [comment]
Would it help if cli randomised the port on supabase start if previous values are taken?

**Comment by jumski** [comment]
@sweatybridge i think it would be a bandaid that would solve this partially but will introduce other new issues. imo any randomization should be avoided if possible.

would it be possible to just allow "port" to be string and parse it as int somewhere upper in the stack?

**Comment by sweatybridge** [comment]
Yes, it should be possible to transform the value type before decoding using the [LoadEnvHook](https://github.com/supabase/cli/blob/develop/pkg/config/decode_hooks.go).

From the DX perspective though, I think it's unnecessarily tedious to change the ports just to have multiple projects running locally.

I will probably end up implementing both so that users can choose to override with env var if they want and leave them as random if they don't care.

**Comment by jumski** [comment]
Yup, i see that point of view. But they will still need to care to read the randomized ports if they want to connect any services.

What is your approach to the randomization - will you randomize the values in config.toml or just assign random ports on startup when port conflicts are discovered, making them drift from ports in the config?

A flags for 'supabase init' that would allow to intiialize project with port overrides is also another idea:

```
supabase init --db-port=xxxx --api-port=yyy
``` 

Maybe allowing to pass ports from envs would enable all of the above in more elegant way?

Just few cents

**Comment by sweatybridge** [comment]
The `supabase status -o env` command prints the connection strings of the running project as env vars. The workflow I have in mind is based on that

```bash
supabase start
supabase status -o env > .env.local
next dev
```

**Comment by jumski** [comment]
Got it! Yes, it makes sense for simple use cases (which are probably 90% of all use cases :-) )

But those are two, overlapping, but different problems IMO, and it would still be valuable to make supabase accept ports from env, as it is de-facto standard nowadays.

Otherwise supabase dictates the ports and one need to build around it and start it first in the stack to learn the ports.

If i can help in any way please hit me up!

**Comment by sweatybridge** [comment]
So I just tested this again and it seems like [LoadEnvHook](https://github.com/supabase/cli/blob/develop/pkg/config/decode_hooks.go) is already working as expected. This is the setup I used

```toml
[db]
port = "env(SECRET_VALUE)"
```

and in `.env.local` file

```
SECRET_VALUE=5432
```

Run `supabase db start` and observe that the exposed host port is correctly set to 5432.

```
$ docker ps -a
CONTAINER ID   IMAGE                                                                    COMMAND                   CREATED              STATUS                        PORTS                                         NAMES
028a79b23bda   public.ecr.aws/supabase/postgres:17.6.1.058                              "sh -c '\ncat <<'EOF'…"   About a minute ago   Up About a minute (healthy)   0.0.0.0:5432->5432/tcp, [::]:5432->5432/tcp   supabase_db_cli
```

I will close this issue as completed but let me know if you run into problems using it.
